src/core/Controls.ts

```diff
@@ -1,24 +1,75 @@
 import * as THREE from "three";
-
+import { Game } from "./Game";
 /**
  * An instance of this class perceives a user's actions within the game,
  * and controls the game input.
  */
 export class Controls {
-    private camera: THREE.PerspectiveCamera;
+    private static camera: THREE.PerspectiveCamera;
+    public static controls: any;
+    public static euler = new THREE.Euler( 0, 0, 0, 'YXZ' );
     public constructor() {
-        this.camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000);
+        Controls.camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000);
+        Controls.createControls()
     }
     public getCamera(): THREE.Camera {
-        return this.camera;
+        return Controls.camera;
     }
     /**
      * Update this controls within the {@link context} specified.
      * @param context a context object to update the controls with
      */
     public update(context: { viewport: { width: number, height: number } }): void {
         const { width, height } = context.viewport;
-        this.camera.aspect = width / height;
-        this.camera.updateProjectionMatrix();
+        Controls.camera.aspect = width / height;
+        Controls.camera.updateProjectionMatrix();
+    }
+    public static createControls(): void {
+        let speed = 10, sensitivity = 1
+        // flyControls for comfortable devloping   
+        let canvas = document.querySelector('#canvas') as HTMLCanvasElement;
+          const canvasClick = () => {
+            canvas.requestPointerLock();
+          };
+          
+          document.addEventListener('pointerlockchange', () => {console.log('sam loh')}, false);
+          canvas.addEventListener('click', canvasClick, false);
+          
+          canvas.addEventListener("mousemove", (event: any) => {
+            console.log('asd')
+            const movementX = event.movementX || event.mozMovementX || event.webkitMovementX || 0;
+            const movementY = event.movementY || event.mozMovementY || event.webkitMovementY || 0;
+        
+            Controls.euler.x -= movementY / (1 / sensitivity * 1000)
+            Controls. euler.y -= movementX / (1 / sensitivity * 1000) 
+        
+            Controls.euler.x = Math.max(Math.min(Math.PI/2, Controls.euler.x), -Math.PI/2)
+            Controls.camera.quaternion.setFromEuler( Controls.euler );
+            Game.needRender = true
+        });
+        window.addEventListener('keydown', (event) => {
+            console.log(event.code)
+            switch (event.code) {
```

Comment:
[nit]
prefer a map `keydowns` with `event.code`s as its keys and the respective `switch`-`case` blocks as its values.

For instance,
```typescript
    private cameraSpeed = 10;
    private keydowns = new Map([
        ['KeyW', () => this.translateCameraTo({ z: -this.cameraSpeed})],
        ['KeyD', () => this.translateCameraTo({ x: this.cameraSpeed})],
        // ...
    ]);
    private translateCameraTo(params: Partial<{ x: number, y: number, z: number }>): void {
        const { x, y, z } = params;
        if (x) this.camera.translateX(x);
        if (y) this.camera.translateY(y);
        // ...
    }
    private _onKeyDown(event: KeyboardEvent): void {
        if (this.keydowns.has(event.code)){
            const processThisEventCode = this.keydowns.get(event.code);
            if (processThisEventCode) processThisEventCode();
        }
        this._fireChangeEvent();
    }
    private _fireChangeEvent(): void {
        // call the callback received in the addListener(callback) method of this class,
        // which would make the Game class's instance to re-render its scene
    }
    public addListener(callback: (...args: any) => any): void {
        // keep reference to the callback function, and call it when a change occurs
    }
```

Reply:
Nice refactoring, will be done